[#8921][feat] Added symetric memory AllReduce strategy

MrGeva opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added symmetric memory all-reduce acceleration for H100+ GPUs to improve distributed training performance
  * Introduced SYMM_MEM strategy option for all-reduce operations
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

<!--
Please write the PR title by following this template:

**[JIRA ticket/NVBugs ID/GitHub issue/None][type] Summary**

Valid ticket formats:
  - JIRA ticket: [TRTLLM-1234] or [FOOBAR-123] for other FOOBAR project
  - NVBugs ID: [https://nvbugs/1234567]
  - GitHub issue: [#1234]
  - No ticket: [None]

Valid types (lowercase): [fix], [feat], [doc], [infra], [chore], etc.

Examples:
  - [TRTLLM-1234][feat] Add new feature
  - [https://nvbugs/1234567][fix] Fix some bugs
  - [#1234][doc] Update documentation
  - [None][chore] Minor clean-up

Alternative (faster) way using CodeRabbit AI:

**[JIRA ticket/NVBugs ID/GitHub issue/None] @coderabbitai title**

NOTE: "@coderabbitai title" will be replaced by the title generated by CodeRabbit AI, that includes the "[type]" and title.
For more info, see /.coderabbit.yaml.

-->

## Description

<!--
Please explain the issue and the solution in short.
-->

## Test Coverage

<!--
Please list clearly what are the relevant test(s) that can safeguard the changes in the PR. This helps us to ensure we have sufficient test coverage for the PR.
-->

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- The reviewers assigned automatically/manually are appropriate for the PR.


- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

MrGeva committed:
Added symm mem strategy

Signed-off-by: Eran Geva <19514940+MrGeva@users.noreply.github.com>

CR fixes

Signed-off-by: Eran Geva <19514940+MrGeva@users.noreply.github.com>

removed sym mem from AUTO flow

Signed-off-by: Eran Geva <19514940+MrGeva@users.noreply.github.com>

fixed process groups handling and fallback

Signed-off-by: Eran Geva <19514940+MrGeva@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #23543](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/23543/) [ run ] triggered by Bot. Commit: `3cb8ff4`

suyoggupta commented on tensorrt_llm/_torch/distributed/ops.py:
```diff
@@ -516,13 +518,17 @@ def __init__(self,
 
-                - AUTO: AUTO chooses between NCCL and MIN_LATENCY mode based on a heuristic policy.
+                - AUTO: AUTO chooses the best available strategy. Will try SYMM_MEM first (if available),
+                  then MNNVL, then choose between NCCL and MIN_LATENCY based on a heuristic policy.
```
Are we sure we want to update the "AUTO" behavior? This can impact the perf of various models in production.. and thus far trtllm hasn't used this? Perhaps a safer / potentially less disruptive way maybe for us to introduce in AutoDeploy first?

tensorrt-cicd commented:
[PR_Github #23584](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/23584/) [ run ] triggered by Bot. Commit: `668b514`

tensorrt-cicd commented:
[PR_Github #23584](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/23584/) [ run ] completed with state `SUCCESS`. Commit: `668b514` 
[/LLM/main/L0_MergeRequest_PR pipeline #17746](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/17746/) completed with status: 'FAILURE'

tensorrt-cicd commented:
[PR_Github #23625](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/23625/) [ run ] triggered by Bot. Commit: `ef29244`

MrGeva commented:
/bot run

MrGeva commented on tensorrt_llm/_torch/auto_deploy/distributed/trtllm.py:
```diff
@@ -26,9 +26,9 @@ def trtllm_allreduce(tensor, op, all_reduce_params=None):
+            # Use SYMM_MEM strategy (tries symmetric memory first, falls back to AUTO if needed)
             _allreduce_cache[cache_key] = AllReduce(
-                mapping=p_config, strategy=AllReduceStrategy.AUTO, dtype=tensor.dtype
+                mapping=p_config, strategy=AllReduceStrategy.SYMM_MEM, dtype=tensor.dtype
```
yes, I will take it

MrGeva committed:
split dist custom ops and use templated patterns

Signed-off-by: Eran Geva <19514940+MrGeva@users.noreply.github.com>

fixed tests, readme, removed rms torch pattern

Signed-off-by: Eran Geva <19514940+MrGeva@users.noreply.github.com>

removed legacy names

Signed-off-by: Eran Geva <19514940+MrGeva@users.noreply.github.com>

simplified collectives.py

Signed-off-by: Eran Geva <19514940+MrGeva@users.noreply.github.com>

removed torch rms alred op

Signed-off-by: Eran Geva <19514940+MrGeva@users.noreply.github.com>

Fixed CR comments

Signed-off-by: Eran Geva <19514940+MrGeva@users.noreply.github.com>

added dist_backend arg

Signed-off-by: Eran Geva <19514940+MrGeva@users.noreply.github.com>

removed try/catch in trtllm)dist.py

Signed-off-by: Eran Geva <19514940+MrGeva@users.noreply.github.com>

MrGeva committed:
fixed rebase, fixed rms norm fusion to use the correct strategy, enhanced rms test to check strategy

Signed-off-by: Eran Geva <19514940+MrGeva@users.noreply.github.com>

MrGeva committed:
fixed pre-commit hook issue

Signed-off-by: Eran Geva <19514940+MrGeva@users.noreply.github.com>

MrGeva committed:
Merge branch 'main' into egeva/sym_mem

Signed-off-by: Eran Geva <19514940+MrGeva@users.noreply.github.com>

MrGeva committed:
removed wrong rebase 

Signed-off-by: Eran Geva <19514940+MrGeva@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #26471](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/26471/) [ run ] completed with state `SUCCESS`. Commit: `51898f4` 
[/LLM/main/L0_MergeRequest_PR pipeline #20123](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/20123/) completed with status: 'FAILURE'

tensorrt-cicd commented:
[PR_Github #26555](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/26555/) [ run ] completed with state `SUCCESS`. Commit: `51898f4` 
[/LLM/main/L0_MergeRequest_PR pipeline #20194](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/20194/) completed with status: 'FAILURE'

MrGeva commented:
/bot run

tensorrt-cicd commented:
[PR_Github #26639](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/26639/) [ run ] triggered by Bot. Commit: `51898f4`

hyukn commented on tensorrt_llm/_torch/distributed/ops.py:
```diff
@@ -674,20 +711,39 @@ def forward(
+                    f"Using SymmetricMemoryAllReduce (MULTIMEM) for input shape {input.shape}"
+                )
+                return symm_mem_output
+        elif self.symm_mem_allreduce and all_reduce_params.fusion_op != AllReduceFusionOp.NONE:
```
If fusion is not supported, does it mean fall back or directly skip this op? If fallback, the other strategy might have uninitialized workspace.

hyukn reviewed (approved):
Since multi-GPU CI has paased, it might be a good idea to merge the PR first. But I am still not sure about the fusion scenarios.

MrGeva committed:
fixed CR comment for logging

Signed-off-by: Eran Geva <19514940+MrGeva@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #27220](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/27220/) [ run ] triggered by Bot. Commit: `c02bdc8`

tensorrt-cicd commented:
[PR_Github #27220](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/27220/) [ run ] completed with state `SUCCESS`. Commit: `c02bdc8` 
[/LLM/main/L0_MergeRequest_PR pipeline #20780](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/20780/) completed with status: 'FAILURE'

tcherckez-nvidia reviewed (approved):
see comments. approved.

MrGeva committed:
fixed CR comments

Signed-off-by: Eran Geva <19514940+MrGeva@users.noreply.github.com>

MrGeva committed:
fixed workspace sizes

Signed-off-by: Eran Geva <19514940+MrGeva@users.noreply.github.com>
